rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -0,0 +1,983 @@
+                    match asset_update_event_telemetry_receiver.shutdown().await {
+                        Ok(()) => {
+                            log::info!("AssetUpdateEventTelemetryReceiver shutdown");
+                        }
+                        Err(e) => {
+                            log::error!("Error shutting down AssetUpdateEventTelemetryReceiver: {e}");
+                            // try shutdown again, but not indefinitely
+                            if shutdown_attempt_count < 3 {
+                                shutdown_attempt_count += 1;
+                                shutdown_notifier.notify_one();
+                            }
+                        }
+                    }
+                  },
+                  aep_msg = asset_endpoint_profile_update_event_telemetry_receiver.recv() => {
+                    if let Some(m) = aep_msg {
+                        match m {
+                            Ok((aep_update_event_telemetry, ack_token)) => {
+                                let Some(aep_name) = aep_update_event_telemetry.topic_tokens.get("ex:aepName") else {
+                                    log::error!("AssetEndpointProfileUpdateEventTelemetry missing ex:aepName topic token.");
+                                    continue;
+                                };
+                                // Try to send the notification to the associated receiver
+                                match aep_update_event_telemetry_dispatcher.dispatch(aep_name, (aep_update_event_telemetry.payload.into(), ack_token)) {
+                                    Ok(()) => {
+                                        log::debug!("AssetEndpointProfileUpdateEventTelemetry dispatched for aep name: {aep_name:?}");
+                                    }
+                                    Err(DispatchError::SendError(payload)) => {
+                                        log::warn!("AssetEndpointProfileUpdateEventTelemetryReceiver has been dropped. Received Telemetry: {payload:?}",);
+                                    }
+                                    Err(DispatchError::NotFound(payload)) => {
+                                        log::warn!("AssetEndpointProfile is not being observed. Received AssetEndpointProfileUpdateEventTelemetry: {payload:?}",);
+                                    }
+                                }
+                            }
+                            Err(e) => {
+                                // This should only happen on errors subscribing, but it's likely not recoverable
+                                log::error!("Error receiving AssetEndpointProfileUpdateEventTelemetry: {e}. Shutting down AssetEndpointProfileUpdateEventTelemetryReceiver.");
+                                // try to shutdown telemetry receiver, but not indefinitely
+                                if shutdown_attempt_count < 3 {
+                                    shutdown_notifier.notify_one();
```

Comment:
this will trigger shutting down both telemetry receivers - is that the desired behavior, or should the other keep running?

Reply:
modified